arangod/Replication2/StateMachines/Document/DocumentStateSnapshot.h

```diff
@@ -23,18 +23,229 @@
+      -> inspection::Status {
+    target = std::string{
+        std::visit(overload{
+                       [](state::Ongoing const&) { return kStringOngoing; },
+                       [](state::Finished const&) { return kStringFinished; },
+                       [](state::Aborted const&) { return kStringAborted; },
+                   },
+                   source)};
+    return {};
+  }
+};
+
+/*
+ * Used to retrieve information about the current state of a snapshot.
+ */
+struct SnapshotStatus {
+  SnapshotState const& state;
+  ShardID shardId{};
+  std::optional<uint64_t> totalDocs{std::nullopt};
+  uint64_t docsSent{0};
+  std::size_t batchesSent{0};
+  std::size_t bytesSent{0};
+  std::chrono::system_clock::time_point startTime{
+      std::chrono::system_clock::now()};
+  std::chrono::system_clock::time_point lastUpdated{startTime};
+  std::optional<std::chrono::system_clock::time_point> lastBatchSent{
+      std::nullopt};
+
+  template<class Inspector>
+  inline friend auto inspect(Inspector& f, SnapshotStatus& s) {
+    return f.object(s).fields(
+        f.field(kStringState, s.state)
+            .transformWith(SnapshotStateTransformer{}),
+        f.field(kStringShardId, s.shardId),
+        f.field(kStringTotalDocsToBeSent, s.totalDocs),
+        f.field(kStringDocsSent, s.docsSent),
+        f.field(kStringTotalBatches, s.batchesSent),
+        f.field(kStringTotalBytes, s.bytesSent),
+        f.field(kStringStartTime, s.startTime)
+            .transformWith(inspection::TimeStampTransformer{}),
+        f.field(kStringLastUpdated, s.lastUpdated)
+            .transformWith(inspection::TimeStampTransformer{}),
+        f.field(kStringLastBatchSent, s.lastBatchSent)
+            .transformWith(inspection::TimeStampTransformer{}));
+  }
+};
+
+/*
+ * Used when we want to retrieve information about all snapshots taken by a
+ * leader.
+ */
+struct AllSnapshotsStatus {
+  std::unordered_map<SnapshotId, SnapshotStatus> snapshots;
+
+  template<class Inspector>
+  inline friend auto inspect(Inspector& f, AllSnapshotsStatus& s) {
+    return f.object(s).fields(f.field(kStringSnapshots, s.snapshots));
+  }
+};
+
+/*
+ * This is what the leader uses to keep track of current snapshots.
+ */
+class Snapshot {
+ public:
+  static inline constexpr std::size_t kBatchSizeLimit{16 * 1024 *
+                                                      1024};  // 16MB
+
+  explicit Snapshot(SnapshotId id, ShardID shardId,
+                    std::unique_ptr<ICollectionReader> reader);
+
+  Snapshot(Snapshot const&) = delete;
+  Snapshot(Snapshot&&) = delete;
+  Snapshot const& operator=(Snapshot const&) = delete;
+  Snapshot const& operator=(Snapshot&&) = delete;
+
+  auto fetch() -> ResultT<SnapshotBatch>;
+  auto finish() -> Result;
+  auto abort() -> Result;
+  [[nodiscard]] auto status() const -> SnapshotStatus;
+
+ private:
+  SnapshotId _id;
+  std::unique_ptr<ICollectionReader> _reader;
+  SnapshotState _state;
+  SnapshotStatus _status;
```

Comment:
I also thought about that myself, but I guess the laziness took the best of me. I would rather rename it in this PR, because otherwise I'm afraid it's going to be used in more and more places.